File: iamminji/leetcode/medium/binary_tree_zigzag_level_order_traversal.py
# 103. Binary Tree Zigzag Level Order Traversal
# https://leetcode.com/problems/binary-tree-zigzag-level-order-traversal/description/

from collections import defaultdict


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def zigzagLevelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """

        cache = defaultdict(list)
        max_depth = -1

        stack = [(root, 0)]
        while stack:
            node, depth = stack.pop()

            if node is not None:
                max_depth = max(max_depth, depth)

                cache[depth].append(node.val)

                stack.append((node.left, depth + 1))
                stack.append((node.right, depth + 1))

        return [cache[depth][::-1] if depth % 2 == 0 else cache[depth] for depth in range(max_depth + 1)]

File: kangtaku/binary_tree_zigzag_level_order_traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        Queue<Dt> q=new LinkedList<>();
        List<List<Integer>> ans=new ArrayList<>();
        q.add(new Dt(0, root));
        while(!q.isEmpty()){
            Dt head=q.poll();
            if(head.node==null)continue;
            if(ans.size()==head.depth){
                ans.add(new ArrayList<>());
            }
            ans.get(head.depth).add(head.node.val);
            q.add(new Dt(head.depth+1,head.node.left));
            q.add(new Dt(head.depth+1,head.node.right));
        }
        for(int i=1;i<ans.size();i+=2){
            Collections.reverse(ans.get(i));
        }
        return ans;
    }
}

class Dt {
    int depth;
    TreeNode node;
    Dt(int depth, TreeNode node) {this.depth=depth;this.node=node;}
}
